# **Bioinformatique**
## **Découverte des commandes de base Unix**

- pwd, ls
- du, df
- tree, cd
- mkdir, touch, echo, nano
- cp, mv
- rm, rmdir
- cat, head, tail
- wc, sort, cut, grep, sed
- md5sum


## Avant d'aller plus loin
---
<mark>Ce notebook doit être dans le dossier `bioinfo1`. Si ce n'est pas le cas, supprimez le du dossier où vous l'avez mis et mettez le de nouveau directement dans le bon dossier depuis votre ordinateur.</mark>

<div class="alert alert-block alert-danger"><b>Attention:</b> 
Ne travaillez pas directement sur ce notebook pour ne pas le perdre. Dupliquez-le et renommez-le par exemple en ajoutant vos initiales et travaillez sur cette nouvelle copie. Pour ce faire, dans le panneau de gauche, faites un clic droit sur le fichier et sélectionnez "Duplicate". Puis, toujours dans la colonne de gauche, faites un clic droit sur cette copie et sélectionnez "rename" pour changer le nom. Ouvrez ensuite cette nouvelle version en double cliquant dessus. Vous êtes prêt(e) à démarrer! <br>
<br>
<b>N'oubliez pas de sauvegarder régulièrement votre notebook</b>: <kbd>Ctrl</kbd> + <kbd>S</kbd>. ou en cliquant sur l'icone 💾 en haut à gauche de votre notebook ou dans le Menu du JupyterLab "File puis "Save Notebook"!
</div>

<div class="alert alert-block alert-info"> 
   
<b>Rappel :</b> Vous pouvez prendre des notes directement dans ce notebook, en ajoutant une cellule Markdown en cliquant sur l'icône <kbd>➕</kbd> dans la barre des menus, et en choisissant son format dans le menu déroulant. 
Les astuces pour utiliser les cellules des notebooks sont rappelées tout en bas dans un cadre bleu ⏬.
</div>

__=> A propos de ce notebook jupyter__

Ce notebook jupyter notebook est en **bash**, ce qui signifie que les commandes que vous entrerez dans les cellules de `Code` seront directement lues en Bash par le serveur. <br> Vous pourriez executez ces mêmes commandes dans un  `Terminal` (l'effroyable fenêtre noire :😰). 

>_Facultatif, si vous êtes curieux de voir un terminal, vous pouvez **transitoirement** en ouvrir un sur adenine:_
>- _dans le menu in the `File` dans la barre du haut, selectionnez `New Launcher` ou cliquez sur le `+` juste en dessous de `File`_
>- _ouvrez un `Terminal` Bash_
>- _Vous pouvez copier et coller les 1ères commandes de ce notebook présentes dans les cellules de`Code` juste après le prompt `$` dans le terminal._ <span style="color:red">
Mais surtout, revenez ensuite au notebook pour tout réaliser dedans, sinon vous perdez tout l'intérêt d'utiliser une interface Jupyterlab!</span>


## 1. Introduction
---

**Bash** est un interpréteur de lignes de commande (*shell*) utilisé sur des systèmes Unix / Linux. C'est un des *shells* les plus utilisés sous Linux. Si vous souhaitez mieux comprendre la différence entre *shell* et terminal, jetez un oeil à cette [video](https://www.youtube.com/watch?v=U2b-MYcSCLc).

- Dans Unix, tous les caractères sont sensibles à la casse.
- N'utilisez pas d'accents et évitez les caractères spéciaux.   
- Dans les cellules de `Code`, les lignes débutant par un `#` sont des commentaires qui ne seront pas interprétées comme des commandes. Elles ont été ajoutées pour vous aider.  
- Vous pouvez ajouter vos propres commentaires ou notes dans ce document, soit dans des cellules de `Code` en utilisant `#`, soit dans des cellules de type `Markdown` que vous pouvez ajouter en cliquant sur le <kbd>➕</kbd> dans le menu en haut à gauche de ce notebook.  
    
**Dans ce notebook, vous aurez 2 types de cellule de code:**
    
- <span style="color:blue">des cellules sans code:</span> vous devrez trouver le code adéquat
- <span style="color:green">des cellules avec le code déjà renseigné:</span> vous devrez comprendre les commandes correspondantes.


<mark>Dans ce TP, vous lancerez chaque cellule l'une après l'autre dans l'ordre. Si vous revenez en arrière et rééxécuter une cellule, veillez à relancer les suivantes. Les cellules sont numérotées dans leur ordre d'éxécution.</mark>

## 2. Système de fichiers 
---

Le système de fichiers (*file system*) d'un système d'exploitation (*operating system* ou *OS*) permet l'organisation hiérarchique des fichiers (*file*) et des répertoires/dossiers (*directory*). Un ensemble de commandes Unix permettent de se déplacer dans cette arborescence, de créer/renommer/déplacer/effacer des fichiers ou des répertoires. 


### 2.1. Répertoire de travail/répertoire courant
---

Pour se déplacer dans l'arborescence, il faut d'abord savoir à quel niveau on se trouve à un instant donné, c'est-à-dire **connaitre le répertoire de travail** (*working directory* ou *wd*) encore appelé **répertoire courant**. 

<span style="color:blue">- Exécutez la commande `pwd` qui affiche le répertoire de travail (<i>print working directory</i>):</span>

In [4]:
# cell 1
pwd

/srv/home/mzidi/introduction_plasma/bioinfo1


<div class="alert alert-block alert-success">
Vous devriez obtenir ceci: /srv/home/votre_login/introduction_plasma/bioinfo1
</div>

Il s'agit du chemin absolu (*absolute path*) du répertoire de travail. 

Remarquez qu' <span style="color:red"><b>un chemin absolu commence toujours par un "slash" `/`</b></span> , qui désigne la racine (*root*) de l'arborescence.

Remarquez que le chemin affiché contient votre `Username`, correspondant à votre `home`,  ainsi que le nom de l'environnement que vous avez chargé dans PlasmaBio. En effet un répertoire est créé dans votre home lorsque vous chargez un environnement dans PlasmaBio pour la 1ère fois.

<div class="alert alert-block alert-warning">
<b>La <code>home</code> est votre répertoire principal</b> sur adenine (et tout autre serveur). Le symbole <code>~</code> (touches <kbd>AltGr</kbd>+<kbd>2</kbd> sur Azerty) est un raccourci de votre home. Vous pouvez utiliser ce symbole plutôt que le chemin absolu/relatif de votre home.</div>

### 2.2. Contenu d'un répertoire
---

#### 2.2.A. Lister le contenu du répertoire de travail
---


<span style="color:blue">- Listez (<i>list</i>) les fichiers/répertoires du répertoire courant à l'aide de la commande `ls`:</span>

In [5]:
# cell 2
ls

bioinfo_1_correction.ipynb  bioinfo_1.ipynb


Comparez le résultat à ce qui est affiché dans la barre latérale gauche de l'interface. Ce doit être identique.

L'affichage se fait sous forme d'une série de valeurs sur une même ligne ou plusieurs lignes.

<span style="color:blue">- En consultant l'aide intégrée avec ` --help`, faites-vous une idée du nombre de possibilités pour changer cet affichage (il devrait y avoir beaucoup de possibilités, plus ou moins intéressantes ou compréhensibles...). </span>

In [6]:
# cell 3
ls --help

Usage: ls [OPTION]... [FILE]...
List information about the FILEs (the current directory by default).
Sort entries alphabetically if none of -cftuvSUX nor --sort is specified.

Mandatory arguments to long options are mandatory for short options too.
  -a, --all                  do not ignore entries starting with .
  -A, --almost-all           do not list implied . and ..
      --author               with -l, print the author of each file
  -b, --escape               print C-style escapes for nongraphic characters
      --block-size=SIZE      scale sizes by SIZE before printing them; e.g.,
                               '--block-size=M' prints sizes in units of
                               1,048,576 bytes; see SIZE format below
  -B, --ignore-backups       do not list implied entries ending with ~
  -c                         with -lt: sort by, and show, ctime (time of last
                               modification of file status information);
                               with -l:

<span style="color:blue"> - Trouvez le/les argument/s à ajouter à la commande `ls` pour :  </span>

<span style="color:blue"> <b>a-</b>
faire apparaître des fichiers/répertoires cachés, c'est-à-dire les fichiers/répertoires dont le nom commence par un point `.`</span>

In [7]:
# cell 4
ls -a ## ou ls --al

.  ..  bioinfo_1_correction.ipynb  bioinfo_1.ipynb  .ipynb_checkpoints


Ces fichiers/répertoires cachés correspondent à des fichiers/répertoires de configuration qui n'ont pas vocation à être modifiés régulièrement, notamment dans le cadre d'une utilisation scientifique. Néanmoins, ils permettent de configurer vos préférences (langues, éditeur de texte, shell à utiliser, etc.), de stocker des historiques, des identifiants à des sites distants, etc.


<span style="color:blue"> <b>b-</b>
n'avoir qu'un seul nom (un seul fichier ou répertoire) par ligne dans un format complet</span>

In [8]:
# cell 5
# deux solutions possibles

## la plus basique:
ls -1 
## une version plus complète:
ls -l 

# Leur s'affiche l'un en dessous-de l'autre

bioinfo_1_correction.ipynb
bioinfo_1.ipynb
total 196
-rw-rw-r-- 1 mzidi mzidi 125501 Jan  7 13:51 bioinfo_1_correction.ipynb
-rw-rw-r-- 1 mzidi mzidi  71574 Jan  6 11:50 bioinfo_1.ipynb


<span style="color:gray"><i>->L'argument `-1` affiche un élément par ligne mais sans information complémentaire.<br>
L'argument `-l` affiche un élément par ligne mais sans information complémentaire.</i></span>


L'affichage devrait se faire ligne par ligne et indiquer, pour chaque entrée du répertoire de travail, s'il s'agit d'un fichier (`-`) ou un répertoire (`d`), les permissions (`rwx`) de lecture/écriture/exécution (*read/write/execute*), le propriétaire, le groupe, la taille et la date de modification.

- On peut combiner des arguments dans une même commande de manière séparée:

In [9]:
# cell 6
ls -l -h

total 196K
-rw-rw-r-- 1 mzidi mzidi 123K Jan  7 13:51 bioinfo_1_correction.ipynb
-rw-rw-r-- 1 mzidi mzidi  70K Jan  6 11:50 bioinfo_1.ipynb


- ou groupée

In [10]:
# cell 7
ls -lh

total 196K
-rw-rw-r-- 1 mzidi mzidi 123K Jan  7 13:51 bioinfo_1_correction.ipynb
-rw-rw-r-- 1 mzidi mzidi  70K Jan  6 11:50 bioinfo_1.ipynb


<span style="color:green">- <b>Question 1 :</b> Qu'est-ce que l'argument `-h` a permis de faire?</span>

<span style="color:gray">-> Par défaut, l'argument `-l` affiche les tailles en octet ce qui n'est pas toujours lisible. L'ajout de l'argument `-h` permet d'afficher les tailles dans un format lisible par un humain en arrondissant dans les unités les plus appropriées (kilo-octets, mégaoctets, ...).</div>

#### 2.2.B. Lister le contenu de n'importe quel répertoire
---


Pour lister le contenu d'un répertoire autre que le répertoire de travail, il suffit d'utiliser la commande `ls` (suivie de ses éventuels arguments) suivie du **chemin absolu ou relatif** de ce répertoire.

<div class="alert alert-block alert-warning"><ul>
    <li>le <b>chemin absolu</b> commence par un <code>/</code> et toute l'arborence depuis la racine (<i>root</i>) et donnée</il>
    <li>le <b>chemin relatif</b> est positionné par rapport à votre répertoire courant et ne démarre jamais pas un slash</li> </ul> </div>

<span style="color:blue"> - Listez le contenu du répertoire `/srv/data/pass-rs-ue3/`. C'est un répertoire situé ailleurs sur le serveur adénine. </span>


In [11]:
# cell 8
ls -l /srv/data/pass-rs-ue3/

total 12
drwxrwxr-x+ 2 1002 1011 4096 Apr 10  2022 brca1
drwxrwxr-x+ 5 1002 1011 4096 Apr 11  2022 data
drwxrwxr-x+ 2 1124 1136 4096 Jul 24  2024 hg38-chr17-1kgenome


<span style="color:green">- <b>Question 2</b>  Quels droits avez-vous sur ces fichiers/répertoires?</span>

<span style="color:gray">-> Contrairement à votre enseignant qui est propriétaire, le résultat de la commande ci-dessus n'indiquera pas votre login, ce qui signifie que vous faites partie du groupe des "autres utilisateurs". Les permissions des "autres utilisateurs" sont données dans les 3 derniers caractères du 1er champ de chaque ligne. Dans le répertoire <vode>/srv/data/pass-rs-ue3/</code> et ses sous-répertoires, vous avez des droits de lecture (w) voire d'éxécution (x) mais pas d'écriture (- au lieu de w). Cela signifie que vous pouvez lire les fichiers mais pas les modifier. Vous ne pouvez pas non plus créer de répertoire ou de fichiers dans ce répertoire.</span>

Ce répertoire contient des données que nous utiliserons notamment pour le microprojet.

<div class="alert alert-block alert-info"><b>Info :</b>A noter que le <b>répertoire courant</b> peut aussi s'indiquer avec son chemin absolu ou par un <code>.</code> ou <code>./</code>. </div>
Ainsi les 3 commandes suivantes retournent le même résultat!

In [12]:
# cell 9
ls

bioinfo_1_correction.ipynb  bioinfo_1.ipynb


In [13]:
# cell 10
ls ./ # ou ls .

bioinfo_1_correction.ipynb  bioinfo_1.ipynb


In [14]:
# cell 11
ls /srv/home/mzidi/introduction_plasma/bioinfo1 ## ici remplacez mzidi par votre login et transformez cette cellule au format code.

bioinfo_1_correction.ipynb  bioinfo_1.ipynb


#### 2.2.C. Espace disque utilisé
---

- **Sur un répertoire**

Afin de connaître l'espace occupé par un répertoire, il faut utiliser la commande `du` pour *disk usage*.
Plusieurs arguments peuvent être utilisés comme indiqué dans l'aide.

In [15]:
# cell 12
du --help

Usage: du [OPTION]... [FILE]...
  or:  du [OPTION]... --files0-from=F
Summarize disk usage of the set of FILEs, recursively for directories.

Mandatory arguments to long options are mandatory for short options too.
  -0, --null            end each output line with NUL, not newline
  -a, --all             write counts for all files, not just directories
      --apparent-size   print apparent sizes, rather than disk usage; although
                          the apparent size is usually smaller, it may be
                          larger due to holes in ('sparse') files, internal
                          fragmentation, indirect blocks, and the like
  -B, --block-size=SIZE  scale sizes by SIZE before printing them; e.g.,
                           '-BM' prints sizes in units of 1,048,576 bytes;
                           see SIZE format below
  -b, --bytes           equivalent to '--apparent-size --block-size=1'
  -c, --total           produce a grand total
  -D, --dereference-args  deref

<span style="color:green">- <b>Question 3</b>  Que vous affiche la commande ci-dessous?</span>

<div class="alert alert-block alert-info"> <b> Info : </b> Pour remonter d'un niveau dans l'arborescence, on utilise l'écriture `../` qui désigne le répertoire "parent" du répertoire de travail.</div>

In [16]:
# cell 13
du -sh ../

12G	../


<span style="color:gray">-> Espace occupé par le repertoire parent (.) = un seul niveau (-s) en mode lisible humain (-h).</span>

La commande ci-après donne un résultat similaire avec également les répertoires cachés

In [17]:
# cell 14
du -hd1 ../

68K	../.jupyter
8.0K	../.conda
364K	../.git
12K	../.local
12G	../data
192K	../Data
4.0K	../.ipython
164K	../.ipynb_checkpoints
400K	../bioinfo1
64K	../example_data
32K	../binder
12G	../


<span style="color:green">- <b>Question 4 :</b>  Qu'apporte l'agument `d2` dans la commande ci-dessous comparé à la commande avec `d1`?</span>

In [18]:
# cell 15
du -hd2 ../

52K	../.jupyter/lab
8.0K	../.jupyter/nbconfig
68K	../.jupyter
8.0K	../.conda
208K	../.git/objects
28K	../.git/refs
8.0K	../.git/info
32K	../.git/logs
56K	../.git/hooks
4.0K	../.git/branches
364K	../.git
8.0K	../.local/share
12K	../.local
11G	../data/meg-m1-a2b
661M	../data/meg-m1-ue1
12G	../data
96K	../Data/.ipynb_checkpoints
192K	../Data
4.0K	../.ipython
164K	../.ipynb_checkpoints
200K	../bioinfo1/.ipynb_checkpoints
400K	../bioinfo1
12K	../example_data/PLINK_example_files
8.0K	../example_data/.ipynb_checkpoints
32K	../example_data/TSI
64K	../example_data
16K	../binder/.ipynb_checkpoints
32K	../binder
12G	../


<span style="color:gray">-> L'argument <code>d</code> permet d'afficher le total (taille) à une profondeur donnée. Avec d1 on a la même profondeur qu'avec <code>-s .</code>: les informations de taille sont données pour le répertoire courant. En indiquant <code>-d2</code>, on a les tailles sur 2 niveaux de profondeur depuis le répertoire courant.</span>

- **Sur toute la machine**

La commande `df` pour *disk free* donne sur chacun des volumes de la machine l'espace libre et l'espace occupé.

In [19]:
# cell 16
df -h

Filesystem      Size  Used Avail Use% Mounted on
overlay         879G  260G  574G  32% /
tmpfs            64M     0   64M   0% /dev
tmpfs           378G     0  378G   0% /sys/fs/cgroup
shm              64M     0   64M   0% /dev/shm
/dev/sda2       879G  260G  574G  32% /sbin/docker-init
/dev/sdb1        29T  1.2T   26T   5% /srv/data
tmpfs           378G     0  378G   0% /proc/acpi
tmpfs           378G     0  378G   0% /proc/scsi
tmpfs           378G     0  378G   0% /sys/firmware


#### 2.2.D. Arborescence
---

Le système de fichiers est une arborescence plus ou moins complexe constituée d'une racine, de branches et de noeuds sur une profondeur variable. On peut visualiser l'arborescence à partir d'un noeud du système de fichiers à l'aide de la commande `tree`.  

- **Visualisez l'arborescence du répertoire parent du répertoire de travail**

In [20]:
# cell 17
tree ../

../
├── Bienvenue-sur-Plasma.ipynb
├── binder
│   ├── apt.txt
│   ├── environment.yml
│   └── postBuild
├── bioinfo1
│   ├── bioinfo_1_correction.ipynb
│   └── bioinfo_1.ipynb
├── data
│   ├── meg-m1-a2b
│   │   ├── qcm
│   │   │   ├── dir1
│   │   │   │   ├── file1
│   │   │   │   ├── file10
│   │   │   │   ├── file11
│   │   │   │   ├── file12
│   │   │   │   ├── file13
│   │   │   │   ├── file14
│   │   │   │   ├── file15
│   │   │   │   ├── file2
│   │   │   │   ├── file3
│   │   │   │   ├── file4
│   │   │   │   ├── file5
│   │   │   │   ├── file6
│   │   │   │   ├── file7
│   │   │   │   ├── file8
│   │   │   │   └── file9
│   │   │   ├── dir2
│   │   │   │   ├── file10
│   │   │   │   ├── file11
│   │   │   │   ├── file12
│   │   │   │   ├── file3
│   │   │   │   ├── file4
│   │   │   │   ├── file5
│   │   │   │   ├── file6
│   │   │   │   ├── file7
│   │   │   │   ├── file8
│   │   │   │   └── file9
│   │   │   ├── dir3
│   │   │   │   ├── file10
│   │   │   │   ├── file11
│   

Parmi les répertoires présents dans votre répertoire de travail, ne faites pas attention à celui appelé `binder`. Il était nécessaire pour la création de l'environnement.

En revanche, vous pouvez noter qu'il existe un répertoire `example_data` contenant lui-même 2 sous-répertoires.

>*Pour info, ils contiennent des fichiers dans des formats très utilisés en génétique: `.ped`, `.map` et`.info`. Il s'agit pour `.ped` d'un fichier de pedigree avec un individu par ligne avec des metadata dans les colonnes 1 à 6, suivi des génotypes des individus pour les marqueurs génétiques listés dans le fichier `.map` ou `.info`. Nous n'aurons pas l'occasion de travailler sur ce type de fichier, ni d'utiliser les logiciels de génétique associés comme le très populaire PLINK. Mais si vous êtes curieux, vous pouvez regarder la documentation de PLINK et le format de ces fichiers à ce lien https://zzz.bwh.harvard.edu/plink/.*

La commande ci-dessus a été utilisée sans argument. Un argument très pratique est `-L` pour afficher le nombre de niveaux souhaités.

In [21]:
# cell 18
tree -L 2 ../ # affichage de 2 niveaux

../
├── Bienvenue-sur-Plasma.ipynb
├── binder
│   ├── apt.txt
│   ├── environment.yml
│   └── postBuild
├── bioinfo1
│   ├── bioinfo_1_correction.ipynb
│   └── bioinfo_1.ipynb
├── data
│   ├── meg-m1-a2b
│   └── meg-m1-ue1
├── Data
│   └── bioinfo-genial-petite.png
├── example_data
│   ├── PLINK_example_files
│   ├── Temperatures.txt
│   ├── test.txt
│   └── TSI
├── example-notebook_R.ipynb
├── LICENSE
├── README.md
└── simplified_notebook.ipynb

9 directories, 13 files


- **Se déplacer dans l'arboresence**

📌 <span style="color:red">Nous sommes actuellement dans le répertoire <code>~/introduction_plasma/bioinfo1</code>.</span>


<div class="alert alert-block alert-warning"> <b> Aide: retour dans le bon répertoire "~/introduction_plasma/bioinfo1" </b><br>En cas de problème sur cette partie 2.2.D, si vous avez besoin de refaire tourner les commandes dans l'ordre des cellules 19 à 23, retournez dans le répertoire de l'environnement avec la commande <br><code>cd ~/introduction_plasma/bioinfo1</code> et vérifiez que vous y êtes avec la commande <code>pwd</code>. Ces commandes sont déjà écrites dans la cellule 'raw' suivante. Vous avez juste besoin de la transformer en cellule de code et de l'exécuter.</div>

Pour se déplacer dans l'arborescence du système de fichiers, c'est-à-dire changer le répertoire de travail, on utilise la commande `cd` (*change directory*).

<div class="alert alert-block alert-info"> <b> Aide: </b><code>cd</code> est une commande silencieuse si tout s'est bien passé, vous n'aurez de message que si le repertoire de destination n'esxiste pas. Aussi, afin de savoir si vous êtes arrivés à bon port, vérifiez-le avec la commande <code>pwd</code>.<br></div>

<span style="color:blue">- Déplacez-vous dans le sous-répertoire `PLINK_example_files` du répertoire `example_data` en utilisant un <i>chemin relatif</i>, puis vérifier que le <b>working directory</b> a bien été modifié. Vous pouvez effectuer plusieurs lignes de code dans la même cellule. </span>

In [22]:
# cell 19
cd ../example_data/PLINK_example_files
pwd

/srv/home/mzidi/introduction_plasma/example_data/PLINK_example_files


<span style="color:blue">- Déplacez-vous dans le sous répertoire `TSI` du répertoire `example_data` en utilisant un <b>chemin relatif</b> et vérifiez que vous vous êtes bien déplacés dans le bon répertoire.

<div class="alert alert-block alert-info"> <b> Astuce: </b><i>Pour remonter d'un niveau dans l'arborescence, on utilise l'écriture <code>../</code> qui désigne le répertoire "parent" du répertoire de travail.</i></div>

In [23]:
# cell 20
cd ../TSI
pwd

/srv/home/mzidi/introduction_plasma/example_data/TSI


<span style="color:blue">- Déplacez vous dans le répertoire correspondant à l'environnement de ce notebook `introduction_plasma` avec le <b>chemin relatif</b>. Vérifiez que vous y êtes bien arrivés.</span>

In [24]:
# cell 21
cd ../../
pwd

/srv/home/mzidi/introduction_plasma


<span style="color:blue">- Déplacez-vous dans le répertoire `TSI` en utilisant le <b>chemin relatif</b>. Vérifiez où vous êtes.</span>

In [25]:
# cell 22
cd ./example_data/TSI
pwd

/srv/home/mzidi/introduction_plasma/example_data/TSI


<span style="color:blue">- Allez à présent dans le répertoire `example_data` en utilisant le <b>chemin absolu</b>. Vérifiez où vous êtes.

<b><i>Astuce:</i></b> <i>Utilisez la flèche de tabulation de votre clavier pour autocompléter le nom des répertoires!</i></span>

In [26]:
# cell 23
cd /srv/home/mzidi/introduction_plasma/example_data
pwd

/srv/home/mzidi/introduction_plasma/example_data


## 3. Manipulation de fichiers et répertoires
---

📌 <span style="color:red">Nous sommes actuellement dans le répertoire <code>~/introduction_plasma/example_data</code>.</span>


<div class="alert alert-block alert-warning"> <b> Aide: retour dans le bon répertoire "~/introduction_plasma/example_data" </b><br>En cas de problème sur cette partie 3, si vous avez besoin de refaire tourner les commandes dans l'ordre des cellules 24 à 39, retournez dans le répertoire de l'environnement avec la commande <br><code>cd ~/introduction_plasma/example_data</code> et vérifiez que vous y êtes avec la commande <code>pwd</code>. Ces commandes sont déjà écrites dans la cellule 'raw' suivante. Vous avez juste besoin de la transformer en cellule de code et de l'exécuter.</div>

### 3.1. Création d'un répertoire et d'un fichier
---

- Pour **créer un répertoire**, on utilise la commande `mkdir` (*make directory*) suivie du chemin du nouveau répertoire et de son nom. 

<span style="color:blue">- Créez un répertoire temporaire `tmp` dans `introduction_plasma` puis déplacez-vous dans ce répertoire et vérifiez que vous y êtes.</span>

In [27]:
# cell 24
mkdir ../tmp
cd ../tmp
pwd

/srv/home/mzidi/introduction_plasma/tmp


- Pour **créer un fichier**, on utilise la commande `touch` suivie du chemin du nouveau fichier avec son nom. 

<span style="color:blue">- Créez un fichier texte `tmp_file.txt` dans le répertoire `tmp` et vérifiez qu'il a bien été créé avec `ls`.</span>

In [28]:
# cell 25
# votre code
touch tmp_file.txt
ls

tmp_file.txt


### 3.2. Copie d'un fichier ou d'un répertoire
---

- Pour **copier un fichier**, on utilise la commande `cp` (*copy*) suivie du chemin (absolu ou relatif) du fichier source, suivi du chemin (absolu ou relatif) du répertoire de copie. 

<span style="color:blue">- Sans vous déplacer, copiez le fichier `test.map` présent dans le répertoire `PLINK_example_files` dans le répertoire `tmp`. NB: il vous faudra réfléchir à la localisation du fichier et du répertoire par rapport au <i>working directory</i>. La tabulation peut de nouveau vous aider à autocompléter les noms des répertoires/fichiers. N'hésitez pas à ajouter une cellule pour vérifier quel est votre répertoire courant si vous l'avez oublié.</span>

In [29]:
# cell 26
# votre code
cp ../example_data/PLINK_example_files/test.map .

<span style="color:blue">-Quelle commande pouvez-vous utiliser pour vérifier que vous avez bien copié le bon fichier au bon endroit ?</span>

In [30]:
# cell 27
# votre code
ls 

test.map  tmp_file.txt


- Pour **copier un répertoire**, la commande est également `cp` en ajoutant l'argument `-r` pour recursif (*recursive*) afin de copier le contenu du répertoire également. Par exemple, dans la cellule suivante, on copie le répertoire `introduction_plasma/example_data` dans le répertoire courant `tmp`.

In [31]:
# cell 28
cp -r ../example_data . 
ls

example_data  test.map  tmp_file.txt


### 3.3 Suppression d'un fichier ou d'un répertoire
---

- Pour **supprimer un fichier**, on utilise la commande `rm` (*remove*) suivie du chemin du fichier à supprimer. 

<span style="color:blue">- Supprimez le fichier `tmp_file.txt` situé dans `tmp` et vérifiez que la suppression a bien eu lieu.</span>


In [32]:
# cell 29
# votre code
rm tmp_file.txt
ls

example_data  test.map


- Pour **supprimer un répertoire** il existe deux méthodes:

**1.** Celle qui fonctionne toujours, **que le répertoire soit vide ou non**,  utilise `rm -r` puis le nom du répertoire. Par exemple, nous pouvons supprimer le répertoire `example_data` que nous avions copié dans `tmp` ainsi:

In [33]:
# cell 30
rm -r example_data
ls

test.map


<div class="alert alert-block alert-danger"><b>Attention:</b> La commande <b>rm -r</b> supprime le répertoire et bien évidemment tout son contenu. Vous pourriez le regretter. Par prudence, la seconde solution est conseillée. </div>

**2. Si le répertoire est vide**, le mieux est d'utiliser `rmdir`. Cette commande ne fonctionne que si le répertoire est vide. Vous pouvez donc le vider avant. C'est ce que nous allons faire en vidant ce qui reste dans le répertoire `tmp` avant de le supprimer.

In [34]:
# cell 31
rm test.map
cd  ..
rmdir tmp
ls

Bienvenue-sur-Plasma.ipynb  Data                      README.md
binder                      example_data              simplified_notebook.ipynb
bioinfo1                    example-notebook_R.ipynb
data                        LICENSE


<div class="alert alert-block alert-info">Avant de passer à la suite de ce notebook on recrée un répertoire temporaire <b>tmp2</b> dans lequel on copie de nouveau le fichier <b>test.map</b>.</div>

In [35]:
# cell 32
mkdir tmp2
cd tmp2
cp ../example_data/PLINK_example_files/test.map .
ls

test.map


### 3.4. Déplacer ou renommer un fichier ou un répertoire avec `mv`
---

- Pour **déplacer (*sans le copier*) un fichier vers un répertoire cible**, on utilise la commande `mv` (*move*) suivie en 1er argument du chemin du fichier à déplacer, puis en 2ème argument du chemin du répertoire cible. 

On déplace par exemple dans le répertoire `introduction_plasma` le fichier `test.map` présent dans le répertoire `tmp2`.

In [36]:
# cell 33
mv ./test.map ../

<span style="color:green">- <b>Question 5</b>  Expliquez pas à pas le détail de la commande ci-dessus.</span>

<span style="color:gray">-> <code>mv</code> sert à déplacer un fichier ou répertoire, <code>./test.map</code> indique qu'on prend le fichier <code>test.map</code> actuellement présent dans le répertoire courant (./). La destination est <code>../</code> ce qui indique qu'il s'agit du répertoire parent (pass_rs_ue3_bistats) du répertoire courant (tmp2).</span>

Une fois déplacé le fichier n'est plus dans le répertoire initial comme on peut s'en assurer ci-dessous.

In [37]:
# cell 34
ls

Vous auriez pu en profiter également pour changer le nom du fichier déplacé en indiquant un nouveau nom après le chemin du répertoire de destination (sans espace). Par exemple on le remet dans `tmp2` en le renommant `test2.map`.

In [38]:
# cell 35
mv ../test.map ./test2.map

<span style="color:green">- <b>Question 6</b>  Expliquez pas à pas le détail de la commande ci-dessus.</span>

<span style="color:gray">-> Cette fois-ci on déplace tout en renommant avec <code>mv</code> le fichier <code>test.map</code> actuellement présent dans le répertoire parent (introduction_plasma du répertoire courant (tmp2). La destination est le répertoire courant (./) en donnant un nouveau nom <code>test2.map</code> au fichier ainsi déplacé. Le fichier source n'est plus présent dans le répertoire source.

- Pour **renommer un fichier**, on utilise en effet également la commande `mv` (*move*) suivie en 1er argument du nom du fichier à renommer, puis en 2ème argument du nouveau nom. C'est donc possible sans changer de répertoire!

Par exemple, on renomme le fichier `test.map` de notre répertoire courant en `example.map`.

In [39]:
# cell 36
mv test2.map example.map
ls

example.map


- Pour **déplacer ou renommer un répertoire** on utilise également la commande `mv`, ce qui déplace le répertoire et tout son contenu. Par exemple, on va renommer le répertoire `tmp2` en `tmp3`.

In [40]:
# cell 37
cd ..
mv tmp2 tmp3

In [41]:
# cell 38
ls

Bienvenue-sur-Plasma.ipynb  Data                      README.md
binder                      example_data              simplified_notebook.ipynb
bioinfo1                    example-notebook_R.ipynb  tmp3
data                        LICENSE


In [42]:
# cell 39
ls tmp3

example.map


## 4. Lire, explorer un fichier et écrire dedans
---

📌 <span style="color:red">Nous sommes actuellement dans le répertoire <code>~/introduction_plasma</code>.</span>


<div class="alert alert-block alert-warning"> <b> Aide: retour dans le bon répertoire "~/introduction_plasma" </b><br>En cas de problème sur cette partie 3, si vous avez besoin de refaire tourner les commandes dans l'ordre des cellules 24 à 39, retournez dans le répertoire de l'environnement avec la commande <br><code>cd ~/introduction_plasma/</code> et vérifiez que vous y êtes avec la commande <code>pwd</code>. Ces commandes sont déjà écrites dans la cellule 'raw' suivante. Vous avez juste besoin de la transformer en cellule de code et de l'exécuter.</div>

### 4.1. Lire un fichier

- Pour **visualiser le contenu d'un fichier** de type texte (on verra dans le projet qu'il existe d'autres formats dits binaires), on utilise la commande `cat` suivie du chemin du fichier à visualiser. 

<span style="color:blue">- Déplacez-vous dans le répertoire `example_data/TSI`.
Visualisez le contenu du fichier `Hapmap3-TSI_13q.21.info` situé dans ce répertoire . </span>

In [43]:
# cell 40
# votre code
cd example_data/TSI
cat Hapmap3-TSI_13q21.info

rs9569366	55232916
rs9597258	55234440
rs4885941	55234854
rs11839457	55243360
rs9591687	55249342
rs9569373	55254118
rs7334154	55259125
rs6561867	55267869
rs9597267	55268861
rs4597206	55274285
rs2997106	55278733
rs3015425	55280381
rs12585364	55286848
rs2997114	55287986
rs2997118	55290109
rs2935461	55302250
rs9569384	55306855
rs1391044	55311160
rs1532453	55315271
rs1391041	55324975
rs10507603	55328914
rs2645938	55329151
rs1117969	55334130
rs2645929	55342530
rs1497292	55348261
rs2645924	55348356
rs2801621	55352569
rs2645920	55357235
rs17052609	55366456
rs2801618	55368150
rs9591697	55368450
rs2801617	55368842
rs17052626	55369150
rs3015430	55371003
rs1995099	55374794
rs975273	55379417
rs1587260	55381246
rs17052641	55384997
rs9537297	55387445
rs9527458	55391080
rs7993035	55397589
rs9597296	55403182
rs1497301	55407531
rs2201715	55409754
rs7992310	55418764
rs11617045	55419026
rs7993082	55419220
rs9597307	55426306
rs1497302	55429744
rs7989771	55434151
rs12430288	55444122
rs377576	55448019
rs2314

Il s'agit d'un fichier au format `.info` d'Haploview, un logiciel d'analyse génétique. Il liste les marqueurs génétiques analysés, chaque ligne décrivant un SNP (*identifiant*, *position physique sur le chromosome*).

- Souvent les fichiers sont très longs et on recommande alors de ne **visualiser** que le **début** ou la **fin** du fichier. Les commandes dédiées à ces tâches sont `head` et `tail`. 

<span style="color:blue">-Exécutez-les sur le fichier `Hapmap3-TSI_13q21.info`.</span>


In [44]:
# cell 41
# votre code
head Hapmap3-TSI_13q21.info

rs9569366	55232916
rs9597258	55234440
rs4885941	55234854
rs11839457	55243360
rs9591687	55249342
rs9569373	55254118
rs7334154	55259125
rs6561867	55267869
rs9597267	55268861
rs4597206	55274285


In [45]:
# cell 42
# votre code
tail Hapmap3-TSI_13q21.info

rs2314014	55448804
rs10467541	55454251
rs573051	55455450
rs12584224	55464336
rs442714	55464860
rs494175	55466299
rs9563369	55469632
rs9597316	55471531
rs7997193	55472364
rs8000414	55472527


Les commandes `head` et `tail` par défaut n'affiche que 10 lignes en bash.

<span style="color:blue">-  Consulter l'aide des commandes `head` et `tail` pour trouver les arguments à utiliser afin de:</span>

In [46]:
# cell 43
# votre code
head --help

Usage: head [OPTION]... [FILE]...
Print the first 10 lines of each FILE to standard output.
With more than one FILE, precede each with a header giving the file name.

With no FILE, or when FILE is -, read standard input.

Mandatory arguments to long options are mandatory for short options too.
  -c, --bytes=[-]NUM       print the first NUM bytes of each file;
                             with the leading '-', print all but the last
                             NUM bytes of each file
  -n, --lines=[-]NUM       print the first NUM lines instead of the first 10;
                             with the leading '-', print all but the last
                             NUM lines of each file
  -q, --quiet, --silent    never print headers giving file names
  -v, --verbose            always print headers giving file names
  -z, --zero-terminated    line delimiter is NUL, not newline
      --help     display this help and exit
      --version  output version information and exit

NUM may have a mu

In [47]:
# cell 44
# votre code
tail --help

Usage: tail [OPTION]... [FILE]...
Print the last 10 lines of each FILE to standard output.
With more than one FILE, precede each with a header giving the file name.

With no FILE, or when FILE is -, read standard input.

Mandatory arguments to long options are mandatory for short options too.
  -c, --bytes=[+]NUM       output the last NUM bytes; or use -c +NUM to
                             output starting with byte NUM of each file
  -f, --follow[={name|descriptor}]
                           output appended data as the file grows;
                             an absent option argument means 'descriptor'
  -F                       same as --follow=name --retry
  -n, --lines=[+]NUM       output the last NUM lines, instead of the last 10;
                             or use -n +NUM to output starting with line NUM
      --max-unchanged-stats=N
                           with --follow=name, reopen a FILE which has not
                             changed size after N (default 5) iterati

<span style="color:blue">   <b>a-</b>
n'afficher que les 5 premières lignes</span>


In [48]:
# cell 45
# votre code
head -n 5 Hapmap3-TSI_13q21.info

rs9569366	55232916
rs9597258	55234440
rs4885941	55234854
rs11839457	55243360
rs9591687	55249342


<span style="color:blue"><b>b-</b>
n'afficher que les 3 dernières lignes</span>


In [49]:
# cell 46
#votre code
tail -n 3 Hapmap3-TSI_13q21.info

rs9597316	55471531
rs7997193	55472364
rs8000414	55472527


<span style="color:blue"><b>c-</b>
afficher tout le contenu sauf les 50 premières lignes</span>


In [50]:
# cell 47
# votre code
tail -n +51 Hapmap3-TSI_13q21.info

rs12430288	55444122
rs377576	55448019
rs2314014	55448804
rs10467541	55454251
rs573051	55455450
rs12584224	55464336
rs442714	55464860
rs494175	55466299
rs9563369	55469632
rs9597316	55471531
rs7997193	55472364
rs8000414	55472527


<span style="color:blue"><b>d-</b>
afficher tout le contenu sauf les 60 dernières lignes</span>


In [51]:
# cell 48
# votre code
head -n -60 Hapmap3-TSI_13q21.info

rs9569366	55232916
rs9597258	55234440


### 4.2. Compter les éléments d'un fichier


- On peut **compter le nombre de caratères, mots, ou lignes** (mais aussi de caractères et de mots) d'un fichier texte à l'aide de la commande `wc`.

Avec son argument `-l` (pour *line*), vous comptez le nombre de lignes.

<span style="color:blue">- Comptez le nombre de lignes du fichier `Hapmap3-TSI_13q21.info`</span>


In [52]:
# cell 49
# votre code
wc -l Hapmap3-TSI_13q21.info

62 Hapmap3-TSI_13q21.info


Avec d'autres arguments, vous comptez d'autres éléments.

In [53]:
# cell 50
wc --help

Usage: wc [OPTION]... [FILE]...
  or:  wc [OPTION]... --files0-from=F
Print newline, word, and byte counts for each FILE, and a total line if
more than one FILE is specified.  A word is a non-zero-length sequence of
characters delimited by white space.

With no FILE, or when FILE is -, read standard input.

The options below may be used to select which counts are printed, always in
the following order: newline, word, character, byte, maximum line length.
  -c, --bytes            print the byte counts
  -m, --chars            print the character counts
  -l, --lines            print the newline counts
      --files0-from=F    read input from the files specified by
                           NUL-terminated names in file F;
                           If F is - then read names from standard input
  -L, --max-line-length  print the maximum display width
  -w, --words            print the word counts
      --help     display this help and exit
      --version  output version information and 

In [54]:
# cell 51
wc -w Hapmap3-TSI_13q21.info

124 Hapmap3-TSI_13q21.info


In [55]:
# cell 52
wc Hapmap3-TSI_13q21.info

  62  124 1183 Hapmap3-TSI_13q21.info


<span style="color:green">- <b>Question 7</b>  Expliquez les résultat des deux commandes ci-dessus en consultant l'aide de la commande.</span>

<span style="color:gray">-> Dans la dellule 51, l'argument <code>-w</code> permet de compter le nombre de mots. Il y en a 124 ce qui est normal car le fichier avait 62 lignes et 2 mots par ligne (ouvrez le en double cliquant dessus depuis le navigateur du Jupyter Lab pour en être convaincu). Dans la cellule 52, aucun argument n'a  été ajouté. Dans ce cas, l'aide indique que la commande affiche d'abord le nombre de lignes, puis le nombre de mots puis le nombre d'octets (taille).</span>

### 4.3. Triez des lignes d'un fichier

- Avec la commande `sort`, il est possible de **trier les lignes** d'un fichier selon un ordre ascendant, descendant, voire aléatoire selon les valeurs d'une colonne. 

In [56]:
# cell 53
sort --help

Usage: sort [OPTION]... [FILE]...
  or:  sort [OPTION]... --files0-from=F
Write sorted concatenation of all FILE(s) to standard output.

With no FILE, or when FILE is -, read standard input.

Mandatory arguments to long options are mandatory for short options too.
Ordering options:

  -b, --ignore-leading-blanks  ignore leading blanks
  -d, --dictionary-order      consider only blanks and alphanumeric characters
  -f, --ignore-case           fold lower case to upper case characters
  -g, --general-numeric-sort  compare according to general numerical value
  -i, --ignore-nonprinting    consider only printable characters
  -M, --month-sort            compare (unknown) < 'JAN' < ... < 'DEC'
  -h, --human-numeric-sort    compare human readable numbers (e.g., 2K 1G)
  -n, --numeric-sort          compare according to string numerical value
  -R, --random-sort           shuffle, but group identical keys.  See shuf(1)
      --random-source=FILE    get random bytes from FILE
  -r, --reverse    

In [57]:
# cell 54
sort -n -r -k2 Hapmap3-TSI_13q21.info

rs8000414	55472527
rs7997193	55472364
rs9597316	55471531
rs9563369	55469632
rs494175	55466299
rs442714	55464860
rs12584224	55464336
rs573051	55455450
rs10467541	55454251
rs2314014	55448804
rs377576	55448019
rs12430288	55444122
rs7989771	55434151
rs1497302	55429744
rs9597307	55426306
rs7993082	55419220
rs11617045	55419026
rs7992310	55418764
rs2201715	55409754
rs1497301	55407531
rs9597296	55403182
rs7993035	55397589
rs9527458	55391080
rs9537297	55387445
rs17052641	55384997
rs1587260	55381246
rs975273	55379417
rs1995099	55374794
rs3015430	55371003
rs17052626	55369150
rs2801617	55368842
rs9591697	55368450
rs2801618	55368150
rs17052609	55366456
rs2645920	55357235
rs2801621	55352569
rs2645924	55348356
rs1497292	55348261
rs2645929	55342530
rs1117969	55334130
rs2645938	55329151
rs10507603	55328914
rs1391041	55324975
rs1532453	55315271
rs1391044	55311160
rs9569384	55306855
rs2935461	55302250
rs2997118	55290109
rs2997114	55287986
rs12585364	55286848
rs3015425	55280381
rs2997106	55278733
rs459720

<span style="color:green">- <b>Question 8 :</b> Quel tri a été fait dans la commande précédente? Expliquez les 3 arguments. </span>

<span style="color:gray">-> Votre réponse: Le tri est numérique, avec une clé de tri correspondant à la 2e colonne, et dans l'ordre décroissant.</span>


<div class="alert alert-block alert-info"> Dans Unix, on peut enchaîner 2 commandes en utilisant à la suite en utilisant le pipe <b>|</b> (touche <kbd>AltGr</kbd>+<kbd>6</kbd> sur Azerty). La commande située après le pipe utilise comme "input" le résultat "output" de la commande à gauche du pipe.</div> 

In [58]:
# cell 55
sort -n -r -k2 Hapmap3-TSI_13q21.info | head

rs8000414	55472527
rs7997193	55472364
rs9597316	55471531
rs9563369	55469632
rs494175	55466299
rs442714	55464860
rs12584224	55464336
rs573051	55455450
rs10467541	55454251
rs2314014	55448804


Ci-dessus on a donc enchainé le tri et décidé de n'afficher que les 10 1ètres lignes du résultat du tri. On voit que cela correspond à l'ordre inverse des 10 dernières lignes du fichier non trié.

In [59]:
# cell 56
tail Hapmap3-TSI_13q21.info

rs2314014	55448804
rs10467541	55454251
rs573051	55455450
rs12584224	55464336
rs442714	55464860
rs494175	55466299
rs9563369	55469632
rs9597316	55471531
rs7997193	55472364
rs8000414	55472527


### 4.4. Extraire les colonnes d'un fichier

- Lorsque le fichier texte est un **"fichier plat"** (chaque ligne contient le même nombre de colonnes, séparées les unes des autres par un délimiteur), il est possible d'**extraire uniquement certaines colonnes** avec la commande `cut`. 

Pour l'exemple, nous allons regarder le fichier `Hapmap3-TSI_13q.21.ped`. Ce fichier est un fichier de génétique au format `.ped` contenant pour chaque ligne les informations pour un individu donné, avec dans les 6 1ères colonnes des informations d'identifiant sur cet un individu (identifiant de la famille, identifiant de l'individu, identifiant du père, identifiant de la mère, sexe, statut par rapport à la maladie). Les colonnes suivantes contiennent les génotypes pour chacun des marqueurs listé dans le fichier `.info` ou `.map` correspondant.

Affichons d'abord les 1ères lignes.

In [60]:
# cell 57
head Hapmap3-TSI_13q.21.ped

NA20502 NA20502 0 0 2 0 C C G A A G A G C C T T A G G G T C G A C A C C G G T G T G G A T T G A T T T G A A T G G G G G G G C C C C C C A A A A G G T T G G A A G A G G T T A A C C T T C C A A A A C C C C G G A A G G C C C C A A A G G G T C T C G A C C A G C C A G C T C A
NA20504 NA20504 0 0 2 0 C C G A A G A G T C T T A G A G T C G A C A T C G G T G 0 0 G A T T G A C T T G A A T G A G G G G G C C C C C C A A A A G G T T A G A A A A G G T T A A C C C T C C A A A A C C C C A G A A A G C C C C A A A G G G C C C C A A C C G G C C G G T T A A
NA20505 NA20505 0 0 2 0 C C G A A G A G T C T T A G A G T C G A C A T C G G T G T G G A T T G A C T T G A A T G A G G G G G C C C C C C A A A A G G T T A G A A G A G G T T A A C C T T C C A A A A C C C C A G A A G G C C C C A A A A G G C C C C A A C C G G C C G G T T A A
NA20506 NA20506 0 0 2 0 C C G A A G A G T C T T A G A G T C G A C A T C G G T G T G G A T T G A C T G G A A G G A G G G G G C C C C C C A A A A G G T T G G A A A A G G C T A A T C C T 

Chaque ligne commence par un identifiant au format `NAXXXXX` dans cet exemple. Comme 62 marqueurs ont été génotypés, l'affichage de chaque ligne continue en dessous.

Consultons l'aide de la commande `cut` pour connaitre les options possibles. 

In [61]:
# cell 58
cut --help

Usage: cut OPTION... [FILE]...
Print selected parts of lines from each FILE to standard output.

With no FILE, or when FILE is -, read standard input.

Mandatory arguments to long options are mandatory for short options too.
  -b, --bytes=LIST        select only these bytes
  -c, --characters=LIST   select only these characters
  -d, --delimiter=DELIM   use DELIM instead of TAB for field delimiter
  -f, --fields=LIST       select only these fields;  also print any line
                            that contains no delimiter character, unless
                            the -s option is specified
  -n                      (ignored)
      --complement        complement the set of selected bytes, characters
                            or fields
  -s, --only-delimited    do not print lines not containing delimiters
      --output-delimiter=STRING  use STRING as the output delimiter
                            the default is to use the input delimiter
  -z, --zero-terminated    line delimite

Sélectionnons à présent les 6 1ères colonnes du fichier.

In [62]:
# cell 59
cut -d" " -f1-6 Hapmap3-TSI_13q.21.ped | head

NA20502 NA20502 0 0 2 0
NA20504 NA20504 0 0 2 0
NA20505 NA20505 0 0 2 0
NA20506 NA20506 0 0 2 0
NA20508 NA20508 0 0 2 0
NA20509 NA20509 0 0 1 0
NA20510 NA20510 0 0 1 0
NA20512 NA20512 0 0 1 0
NA20515 NA20515 0 0 1 0
NA20516 NA20516 0 0 1 0


<span style="color:green">- <b>Question 9:</b> Expliquez la commande ci-dessus. </span>

<span style="color:gray">-> Votre réponse: Le séparateur étant un expace plutôt qu'une tabulation, on utilise `-d` pour indiquer le type de séparateur, puis on choisit les colonnes 1 à 6 avec l'agrument <code>-f.</code> On idnique ensuite le fichier sur lequel on réalise ces manipulations. Le résultat de cette 1ère commande est envoyé comme input de la commande <code>head</code> grâce au pipe <code>|</code>.</span>

<div class="alert alert-block alert-info"> <b> Information: </b>Vous noterez l'odre dans le fichier `Hapmap3-TSI_13q21.info` n'a pas été modifié, et que toutes les colonnes du fichier `Hapmap3-TSI_13q21.ped` n'ont pas disparu. Les commandes `sort` ou `cut` ont seulement affiché le résultat du tri mais elles n'ont pas écrasé les données dans le fichier.<br>
Comment modifie-t-on un fichier ou écrit-on dedans? C'est ce qui vous sera présenté dans le paragraphe 4.6.</div> 

### 4.5 Extraire un motif d'un fichier

Pour extraire un motif, on utilise la commande `grep`. Cela retourne toutes les lignes d'un fichier texte contenant le motif en question indiqué entre des doubles ou simples guillemets.

- Par exemple, on peut extraire la ligne contenant le SNP `rs9597307` du fichier `Hapmap3-TSI_13q21.info`.

In [63]:
# cell 60
grep "rs9597307" Hapmap3-TSI_13q21.info

rs9597307	55426306


- On peut aussi extraire un motif plus court, par exemple toutes les lignes avec le motif "rs95".

In [64]:
# cell 61
grep "rs95" Hapmap3-TSI_13q21.info

rs9569366	55232916
rs9597258	55234440
rs9591687	55249342
rs9569373	55254118
rs9597267	55268861
rs9569384	55306855
rs9591697	55368450
rs9537297	55387445
rs9527458	55391080
rs9597296	55403182
rs9597307	55426306
rs9563369	55469632
rs9597316	55471531


- Ici on extrait tous les SNPs dont l'identifiant se termine par un "3". Pour ce faire, on utilise l'argument `-P` pour lire en langage Perl une *expression régulière*. Entre les guillements, on ajoute un `.` pour indiquer un caratère, n'importe lequel. Le fait d'ajouter `*` permet d'appliquer le `.` à un nombre variable de caractères (un ou plusieurs). Ensuite, comme il y a 2 colonnes, on ajoute `\\t` pour indiquer qu'il y a une tabulation comme séparateur de colonne.

In [65]:
# cell 62
grep -P "rs.*3\\t" Hapmap3-TSI_13q21.info

rs9569373	55254118
rs1532453	55315271
rs10507603	55328914
rs975273	55379417
rs7997193	55472364


- Au dessus, on savait que le rs était dans la 1ère colonne et en début de ligne, mais ce n'est pas toujours le cas. On peut forcer à chercher en début de ligne avec le `^` placé au début du motif recherché.


In [66]:
# cell 63
grep -P "^rs.*3\\t" Hapmap3-TSI_13q21.info

rs9569373	55254118
rs1532453	55315271
rs10507603	55328914
rs975273	55379417
rs7997193	55472364


- On peut aussi chercher dans la 2ème colonne (après la tabulation) tout ce qui commence par "554".

In [67]:
# cell 64
grep -P "\\t554" Hapmap3-TSI_13q21.info

rs9597296	55403182
rs1497301	55407531
rs2201715	55409754
rs7992310	55418764
rs11617045	55419026
rs7993082	55419220
rs9597307	55426306
rs1497302	55429744
rs7989771	55434151
rs12430288	55444122
rs377576	55448019
rs2314014	55448804
rs10467541	55454251
rs573051	55455450
rs12584224	55464336
rs442714	55464860
rs494175	55466299
rs9563369	55469632
rs9597316	55471531
rs7997193	55472364
rs8000414	55472527


- Dans ce dernier  exemple, on recherche toutes les lignes qui se terminent par "1" en utilisant `$` pour indiquer la fin de la ligne.

In [68]:
# cell 65
grep "1$" Hapmap3-TSI_13q21.info

rs9597267	55268861
rs3015425	55280381
rs1532453	55315271
rs2645938	55329151
rs1497292	55348261
rs1497301	55407531
rs7989771	55434151
rs10467541	55454251
rs9597316	55471531


### 4.6. Ecrire dans un fichier

Pour **écrire dans un fichier :**

**1. Sur l'interface Jupyter Lab**, il est possible d'ouvrir le fichier par un double clic sur son nom dans le navigateur à gauche. Si vous le faites, vous verrez qu'il est vide. Vous pouvez y écrire du texte directement. En le fermant, Jupyter Lab vous demandera de le sauvegarder.

**2. Sans interface Jupyter Lab**, comment écrit-on dans un fichier texte?
Plusieurs outils existent, plus ou moins faciles.

Une solution est de **rediriger le résultat d'une commande dans un fichier**. Cela se fait avec avec `>` qui écrase le fichier ou avec `>>` qui écrit à la suite d'un fichier existant.
    
Pour l'exemple, une façon simple d'écrire du texte est d'utiliser la commande `echo` qui écrit du texte sur le terminal et dont vous pouvez rediriger  le contenu dans un fichier.


In [87]:
#cd bioinfo1
mkdir ../tmp3

In [88]:
# cell 66
cd ../tmp3
echo "quelques mots dans mon fichier" > tmp_file.txt

Si vous voulez continuer d'écrire à la suite, il faudra utiliser `>>` plutôt que `>` pour ne pas écraser le contenu.

In [89]:
# cell 67
echo "d'autres mots dans mon fichier" >> tmp_file.txt

Regardez le résultat:

In [90]:
# cell 68
cat tmp_file.txt

quelques mots dans mon fichier
d'autres mots dans mon fichier


> ***Facultatif*** *Pour les plus audacieux, vous pouvez tester le programme `nano`.*
> On veut créer un nouveau fichier de type texte nommé `gene_list.txt` dans le répertoire `tmp` et ayant le contenu suivant (NB: vous séparerez les colonnes par des tabulations):
>
> | symbol | name |
> | --- | --- |
> | APOE | apolipoprotein  |
> | AR | androgen receptor |
> | GRN | granulin |
> | HTT | huntingtin |
> | PSEN1 | presenilin 1 |
>
> <div class="alert alert-block alert-danger">
><b>Attention</b>: 
>    nano est un outil interactif qui doit être appelé depuis un <b>Terminal</b> et pas depuis un Notebook Jupyter. 
></div>
>- Cliquez en haut à gauche sur le "+" et ouvrez un Terminal.
>
> - Déplacez-vous dans le répertoire `tmp` avec la commande `cd example_data/tmp`
> - Executez la commande: `nano gene_list.txt`
>
> **Attention**: si nano s'utilise comme n'importe quel éditeur de texte classique pour la saisie du texte, les raccourcis clavier (pour copier, couper, coller, sauvegarder, quitter) sont inhabituels:
> - <kbd>ALT</kbd>+<kbd>SHIFT</kbd>+<kbd>6</kbd>: copier un texte sélectionné (et le placer dans le presse-papier)
> - <kbd>CTRL</kbd>+<kbd>K</kbd> : couper un texte sélectionné (et le placer dans le presse-papier)
> - <kbd>CTRL</kbd>+<kbd>U</kbd> : coller le texte présent dans le presse-paprier
> - <kbd>CTRL</kbd>+<kbd>O</kbd> : enregistrer le fichier
> - <kbd>CTRL</kbd>+<kbd>X</kbd> : quitter nano




## 5. Sommes de contrôle
---

Les sommes de contrôle sont la signature ou l'empreinte numérique (*hash*) d'un fichier. Elles sont particulièrement utiles pour vérifier l'intégrité d'un fichier. En passant un fichier dans un algorithme particulier, on obtient une somme de contrôle, souvent représentée comme une suite de chiffres et de lettres.

**MD5:** L'agorithme *Message Digest 5* (MD5) est un des algorithmes les plus utilisés et un des plus anciens. Il date de 1991.

Cet algorithme est (était) utilisé en cryptographie. L'empreinte est obtenue est représentée sur 32 caractatères. Par exemple : 
```
0648b3abc5c0e0e7cbf78c466670affa
```

La probabilité que 2 fichiers différents possèdent la même empreinte est de l'ordre de $2^{-128}$.

>*Remarque :* avec un système d'exploitation Mac OS X, l'équivalent de la commande `md5sum` est `md5`


Voici une ***démonstration:***

1. Faisons une copie du fichier dans `tmp_file.txt`.

In [91]:
# cell 69
cp tmp_file.txt tmp_file2.txt

2. On calcule les empreintes MD5 des deux fichiers avec la commande suivante:

In [92]:
# cell 70
md5sum tmp_file.txt tmp_file2.txt

c0e4d8e35c33fff15162af498635086b  tmp_file.txt
c0e4d8e35c33fff15162af498635086b  tmp_file2.txt




=> On voit que les empreintes sont les mêmes donc les fichiers sont identiques (copie l'un de l'autre).

3. Modifions le fichier `tmp_file2.txt` en changeant un seul caractère.

On peut modifier le fichier `tmp_file2.txt`, en l'ouvrant dans le lab, avec nano ou directement en ligne de commande, par exemple, en remplacant un seul caractère (le `a` de `avec` est remplacé par `A`) :

In [93]:
# cell 71
sed -i 's/a/A/' tmp_file2.txt

4. On calcule à nouveau les empreintes MD5 des deux fichiers.

In [94]:
# cell 72
md5sum tmp_file.txt tmp_file2.txt

c0e4d8e35c33fff15162af498635086b  tmp_file.txt
614be7730d995df76a4feeb59761e657  tmp_file2.txt


L'empreinte du fichier `tmp_file.txt` est la même que précédemment car le fichier n'a pas été modifié. Par contre, l'empreinte du fichier `tmp_file2.txt` est complètement différente, alors qu'un seul caractère a été modifié (*a* minuscule remplacé par un *A* majuscule).

Comme nous ne nous en servirons pas par la suite, nous supprimon le reperoire tmp3.

In [95]:
# cell 73
cd ..
rm -r tmp3

In [96]:
# cell 74
ls

Bienvenue-sur-Plasma.ipynb  Data                      README.md
binder                      example_data              simplified_notebook.ipynb
bioinfo1                    example-notebook_R.ipynb  tmp_file.txt
data                        LICENSE


## Conclusion/Perspectives
---

Le *shell* Bash vous donne accès aux commandes Unix (`ls`, `cat`, `pwd`...).

Bash est aussi un langage de programmation que l'ont peututiliser pour automatiser un certain nombre de manipulations sur des fichiers.
En programmation, on a besoin de faire 4 actions essentielles :

1. Stocker de l'information (les variables).
2. Manipuler cette information (les opérateurs).
3. Répéter des actions (les boucles).
4. Prendre des décisions (les tests).

Une fois qu'on sait faire ces 4 actions, on peut pratiquement tout faire.
Nous n'avons pas le temps dans le cadre de cette UE de vous initier à la programmation Bash. Nous ferons une petite initiation avec le langage R dans la partie "Biostatistiques" de l'UE.

<div class="alert alert-block alert-success"><b>
    Bravo! Vous avez accompli vos 1ers pas dans l'univers des commandes Unix avec  succès!</b> <br>
    Vous pourrez désormais vous lancer dans le microprojet sur les variants identifiés par le projet 1000 genomes.<br><br>
    N'oubliez pas de sauvegarder votre rapport. Vous pouvez en faire un export en format html : <br>
- Ouvrez "File" dans le Menu<br>
- Selectionnez "Export Notebook As"<br>
- Exportez votre notebook en HTML<br>
- Vous pouvez maintenant l'ouvrir dans un navigateur sans être connecté à Plasma ! 
</div>

---
---

🙏 Remerciements à Bruno Toupance et Pierre Poulain leur matériel de cours.

<div class="alert alert-block alert-info"> 
    
<b>Rappel</b>, dans un notebook :

- la combinaison de touches <kbd>Ctrl</kbd>+<kbd>Entrée</kbd> exécute une cellule.<br>
- la combinaison de touches <kbd>Shift</kbd>+<kbd>Entrée</kbd> exécute une cellule puis passe à la suivante. C'est équivalent à cliquer sur l'icone ▶️ dans la barre de menu du notebook.<br>
- la combinaison de touches <kbd>Alt</kbd>+<kbd>Entrée</kbd> exécute une cellule puis en crée une nouvelle (vide) en dessous.<br>

Pour ajouter une cellule, vous pouvez aussi cliquer sur l'icone ➕ dans la barre de menu du notebook.
- vous pouvez déplacer les cellules en les glissant pour les réorganiser les unes en-dessous des autres.<br>
- vous pouvez ajouter des commentaires, soit en commençant la ligne par un "#" dans une cellule de code (ces lignes ne seront pas éxécutées), soit dans une nouvelle céllule de type Markdown.<br>
- vous sélectionner le type de cellule dans le menu en haut de votre Vnotebbok.<br>
    - "Code" pour entrer des lignes de commande à éxécuter <br>
    - "Markdown pour ajouter du texte balisé qui peut être formatté<br>
- pour modifier une cellule ̀Markdown, double-cliquez dessus<br>
- attention à la casse des caractères pour les cellules de code<br>
- évitez les caractères spéciaux dans les cellules de code<br>
    
<em>  
To make nice html reports with markdown: <a href="https://dillinger.io/" title="dillinger.io">html visualization tool 1</a> or <a href="https://stackedit.io/app#" title="stackedit.io">html visualization tool 2</a>, <a href="https://www.tablesgenerator.com/markdown_tables" title="tablesgenerator.com">to draw nice tables</a>, and the <a href="https://medium.com/analytics-vidhya/the-ultimate-markdown-guide-for-jupyter-notebook-d5e5abf728fd" title="Ultimate guide">Ultimate guide</a>. <br>
Further reading on JupyterLab notebooks: <a href="https://jupyterlab.readthedocs.io/en/latest/user/notebook.html" title="Jupyter Lab">Jupyter Lab documentation</a>.<br>
Here we are using JupyterLab interface implemented as part of the <a href="https://plasmabio.org/" title="plasmabio.org">Plasmabio</a> project led by Sandrine Caburet, Pierre Poulain and Claire Vandiedonck.
</em>    
</div>

*Last updated on 02/01/2026 by Marwa ZIDI. This notebook was created by @CVandiedonck.*